File: baekjoon/src/dynamic1/No10844.java
package dynamic1;

import java.io.*;
import java.util.StringTokenizer;

/**
 * https://www.acmicpc.net/problem/10844
 */
public class No10844 {
    public static final BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
    public static final BufferedWriter writer = new BufferedWriter(new OutputStreamWriter(System.out));
    public static StringTokenizer tokenizer;

    public static final int MOD = 1000000000;
    public static Integer[][] dp;

    public static int getStairs(int digit, int index) {
        if (digit == 1)
            return dp[digit][index];

        if (dp[digit][index] == null) {
            if (index == 0)
                dp[digit][index] = getStairs(digit - 1, 1) % MOD;
            else if (index == 9)
                dp[digit][index] = getStairs(digit - 1, 8) % MOD;
            else
                dp[digit][index] = (getStairs(digit - 1, index - 1) + getStairs(digit - 1, index + 1)) % MOD;
        }

        return dp[digit][index];
    }

    public static void main(String[] args) throws IOException {
        int digit = Integer.parseInt(reader.readLine());
        dp = new Integer[digit + 1][10];
        for (int i = 0; i < 10; i++)
            dp[1][i] = 1;

        long count = 0;
        for (int i = 1; i < 10; i++)
            count += getStairs(digit, i);

        writer.write(String.valueOf(count % MOD));
        writer.flush();
    }
}

File: baekjoon-classification/src/dynamic/No10844.kt
package dynamic

import java.io.BufferedReader
import java.io.BufferedWriter
import java.io.InputStreamReader
import java.io.OutputStreamWriter

/**
 * https://www.acmicpc.net/problem/10844
 */
fun main() = with(BufferedReader(InputStreamReader(System.`in`))) {
    with(BufferedWriter(OutputStreamWriter(System.out))) {
        val number = readLine().toInt()
        val stairs = Array(number + 1) { i ->
            when (i) {
                0 -> LongArray(10)
                1 -> LongArray(10) { 1 }
                else -> LongArray(10) { -1 }
            }
        }

        for (i in 2..number) {
            for (j in 0..9) {
                when (j) {
                    0 -> stairs[i][j] = stairs[i - 1][1] % 1000000000
                    9 -> stairs[i][j] = stairs[i - 1][8] % 1000000000
                    else -> stairs[i][j] = (stairs[i - 1][j - 1] + stairs[i - 1][j + 1]) % 1000000000
                }
            }
        }

        val sum = (stairs[number].sum() - stairs[number][0]) % 1000000000
        write(sum.toString())
        flush()
    }
}